src/main/scala/chisel3/util/Bitwise.scala

```diff
@@ -80,6 +80,31 @@ object PopCount {
 
   private def _apply_impl(in: Iterable[Bool])(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): UInt =
     SeqUtils.count(in.toSeq)
+
+  /** a threshold PopCount which has optimization to 0,1,2.
+    * @param n the threshold which will be counted
+    * @param x UInt need to be counted
+    * @return true.B when x has more then n set bits.
+    */
+  def atLeast(n: Int, x: UInt): Bool = macro SourceInfoTransform.nxArg
+
+  /** @group SourceInfoTransformMacro */
+  def do_atLeast(n: Int, x: UInt)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): UInt = {
+    def two(x: UInt): (Bool, Bool) = x.getWidth match {
+      case 1 => (x.asBool, false.B)
+      case _ =>
+        val half = x.getWidth / 2
+        val (leftOne, leftTwo) = two(x(half - 1, 0))
+        val (rightOne, rightTwo) = two(x(x.getWidth - 1, half))
+        (leftOne || rightOne, leftTwo || rightTwo || (leftOne && rightOne))
+    }
+    n match {
+      case 0 => true.B
+      case 1 => x.orR
+      case 2 => two(x)._2
+      case 3 => PopCount(x) >= n.U
```

Comment:
This should be `case _`, not `case 3`.  (Otherwise it'll crash for n >= 4.  It's a bug in rocket-chip, too)